api/tools/src/main/java/org/pmiops/workbench/tools/BackfillBillingProjectUsers.java

```diff
@@ -0,0 +1,166 @@
+    WorkspacesApi api = new WorkspacesApi();
+    api.setApiClient(newApiClient(apiUrl));
+    return api;
+  }
+
+  private static BillingApi newBillingApi(String apiUrl) throws IOException {
+    BillingApi api = new BillingApi();
+    api.setApiClient(newApiClient(apiUrl));
+    return api;
+  }
+
+  private static void dryLog(boolean dryRun, String msg) {
+    String prefix = "";
+    if (dryRun) {
+      prefix = "[DRY RUN] Would have... ";
+    }
+    log.info(prefix + msg);
+  }
+
+  /**
+   * Swagger Java codegen does not handle the WorkspaceACL model correctly; it returns a GSON map
+   * instead. Run this through a typed Gson conversion process to coerce it into the desired type.
+   */
+  private static Map<String, WorkspaceAccessEntry> extractAclResponse(WorkspaceACL aclResp) {
+    Type accessEntryType = new TypeToken<Map<String, WorkspaceAccessEntry>>() {}.getType();
+    Gson gson = new Gson();
+    return gson.fromJson(gson.toJson(aclResp.getAcl(), accessEntryType), accessEntryType);
+  }
+
+  private static void backfill(
+      WorkspacesApi workspacesApi,
+      BillingApi billingApi,
+      String billingProjectPrefix,
+      boolean dryRun)
+      throws ApiException {
+    int userUpgrades = 0;
+    for (WorkspaceResponse resp : workspacesApi.listWorkspaces()) {
+      Workspace w = resp.getWorkspace();
+      if (!w.getNamespace().startsWith(billingProjectPrefix)) {
+        continue;
+      }
+
+      String id = w.getNamespace() + "/" + w.getName();
+      if (!"PROJECT_OWNER".equals(resp.getAccessLevel())) {
+        log.warning(
+            String.format(
+                "service account has '%s' access to workspace '%s'; skipping",
+                resp.getAccessLevel(), id));
+        continue;
+      }
+
+      Map<String, WorkspaceAccessEntry> acl =
+          extractAclResponse(workspacesApi.getWorkspaceAcl(w.getNamespace(), w.getName()));
+      for (String user : acl.keySet()) {
+        WorkspaceAccessEntry entry = acl.get(user);
+        if (!"OWNER".equals(entry.getAccessLevel())) {
+          // Only owners should be granted billing project user.
+          continue;
+        }
+        dryLog(
+            dryRun,
+            String.format(
+                "granting billing project user on '%s' to '%s' (%s)",
+                w.getNamespace(), user, entry.getAccessLevel()));
+        if (!dryRun) {
+          billingApi.addUserToBillingProject(w.getNamespace(), "user", user);
+        }
+        userUpgrades++;
+      }
+    }
+
+    dryLog(dryRun, String.format("added %d users as billing project users", userUpgrades));
+  }
+
+  @Bean
+  public CommandLineRunner run() {
+    return (args) -> {
+      CommandLine opts = new DefaultParser().parse(options, args);
+      backfill(
+          newWorkspacesApi(opts.getOptionValue(fcBaseUrlOpt.getLongOpt())),
+          newBillingApi(opts.getOptionValue(fcBaseUrlOpt.getLongOpt())),
```

Comment:
I think you can grab the fcBaseUrl through the WorkbenchConfigProvider if you want to get it through DI instead of command line args.